src/Diagrams/TwoD/Layout/Constrained.hs

```diff
@@ -0,0 +1,671 @@
+-- >   along (direction (1 ^& (-1))) (map centerOf sqs)
+--
+-- <<diagrams/src_Diagrams_TwoD_Layout_Constrained_diagonalLayout.svg#diagram=diagonalLayout&width=400>>
+--
+-- Take a look at the implementations of combinators such as 'sameX',
+-- 'allSame', 'constrainDir', and 'along' for ideas on implementing
+-- your own constraint combinators.
+--
+-- Ideas for future versions of this module:
+--
+-- * Introduce z-index constraints.  Right now the diagrams are just
+--   drawn in the order that they are introduced.
+--
+-- * A way to specify default values --- /i.e./ be able to introduce
+--   new point or scalar variables with a specified default value
+--   (instead of just defaulting to the origin or to 1).
+--
+-- * Doing something more reasonable than crashing for overconstrained
+--   systems.
+--
+-- I am also open to other suggestions and/or pull requests!
+-----------------------------------------------------------------------------
+
+module Diagrams.TwoD.Layout.Constrained
+       ( -- * Basic types
+         Expr, mkExpr, Constrained
+
+         -- * Layout
+       , layout
+
+         -- * Creating constrainable things
+
+         -- | Diagrams, points, /etc./ which will participate in a
+         --   system of constraints must first be explicitly
+         --   introduced using one of the functions in this section.
+       , newDia, newDias, newScalableDia
+       , newPoint, newPointOn
+       , newScalar
+
+         -- * Diagram accessors
+
+         -- | Combinators for extracting constrainable attributes of
+         --   an introduced diagram.
+       , centerOf, xOf, yOf, scaleOf
+
+         -- * Constraints
+       , (====), (=.=), (=^=)
+       , sameX, sameY
+       , allSame
+       , constrainWith
+       , constrainDir
+       , along
+
+       )
+       where
+
+import           Control.Lens         (makeLenses)
+import qualified Control.Lens         as L
+import qualified Control.Lens.Extras  as L
+import           Control.Monad.Except
+import           Control.Monad.State
+import qualified Data.Foldable        as F
+import           Data.Hashable
+import           Data.List            (sortBy)
+import qualified Data.Map             as M
+import           Data.Maybe           (fromJust)
+import           Data.Ord             (comparing)
+import           GHC.Generics
+
+import qualified Math.MFSolve         as MFS
+
+import           Diagrams.Coordinates
+import           Diagrams.Prelude
+
+------------------------------------------------------------
+-- Variables and expressions
+------------------------------------------------------------
+
+-- | An abstract type representing unique IDs for diagrams.  The
+--   constructor is not exported, so the only way to obtain a 'DiaID'
+--   is by calling 'newDia' or 'newDias'. The phantom type parameter
+--   's' ensures that such 'DiaID's can only be used with the
+--   constrained system in which they were introduced.
+newtype DiaID s = DiaID Int
```

Comment:
This (as well as `ConstrainedState`) isn't exported, so you can't see the documentation in haddock. Of course you don't want to export the constructors but I think the documentation is helpful.